jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java

```diff
@@ -1300,7 +1306,12 @@ public void failed(Throwable x)
                 httpChannel.lockedStreamSendCompleted(false);
             }
             if (callback != null)
-                httpChannel._serializedInvoker.run(() -> callback.failed(x));
+            {
+                // Do not use SerializedInvoker here because application code running within
+                // SerializedInvoker could issue blocking write() or close() that would deadlock.
+                // We must execute in case of recursive write within the callback.
+                getRequest().getContext().execute(() -> callback.failed(x));
```

Comment:
@sbordet I don't like this, but we have a test which calls write from within the callback which results in recursion and a StackOverflowError if we execute the callback directly.